anvil/sample.py

```diff
@@ -25,6 +26,165 @@ def sample(loaded_model, action, target_length: int, n_large=20000) -> torch.Ten
     current proposal. A uniform random number, u, is drawn and if u <= A then
     the proposed state phi^j is accepted (and becomes phi^i for the next update)
 
+    Parameters
+    ----------
+    loaded_model: Module
+        loaded_model which is going to be used to generate sample states
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+    batch_size: int
+        the number of states to generate from the loaded_model
+    state_i: torch.Tensor or None
+        the current state of the chain. None if this is the first batch
+
+    Returns
+    -------
+    phi[chain_indices, :]: torch.Tensor
+        chain of configurations generated by the MH algorithm
+    history: torch.BoolTensor
+        boolean tensor containing accept/reject history of chain
+    state: torch.Tensor
+        the current phi state, for continuity between batches
+    """
+    with torch.no_grad():  # don't track gradients
+        z = torch.randn((batch_size, loaded_model.size_in))  # random z configurations
+        phi = loaded_model.inverse_map(z)  # map using trained loaded_model to phi
+        if state_i is not None:
+            phi[0] = state_i
+        log_ptilde = loaded_model(phi)
+    history = torch.zeros(batch_size, dtype=torch.bool)  # accept/reject history
+    chain_indices = torch.zeros(batch_size, dtype=torch.long)
+
+    log_ratio = log_ptilde + action(phi)
+    if not isfinite(exp(float(min(log_ratio) - max(log_ratio)))):
+        raise ValueError("could run into nans")
+
+    i = 0
+    log_ratio_i = log_ratio[0]
+    for j in range(1, batch_size):
+        condition = min(1, exp(float(log_ratio_i - log_ratio[j])))
+        if np.random.uniform() <= condition:
+            chain_indices[j] = j
+            log_ratio_i = log_ratio[j]
+            state_i = phi[j]
+            i = j
+            history[j] = True  # accepted -> True
+        else:
+            chain_indices[j] = i
+
+    return phi[chain_indices, :], history, state_i
+
+
+def thermalisation(loaded_model, action):
+    r"""
+    A (hopefully) short initial sampling phase to allow the system to thermalise.
+
+    Parameters
+    ----------
+    loaded_model: Module
+        loaded_model which is going to be used to generate sample states
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+
+    Returns
+    -------
+    state: torch.Tensor
+        the current phi state, for continuity between batches
+    log_ratio: float
+        the ratio associated with state
+    """
+    t_therm = 1000 # ideally come up with a way of working this out on the fly
+
+    state = sample_batch(
```

Comment:
if not using values from an unpack I think it's cleaner to put

`_, _, state = sample_batch(..)`

if somebody reads your code they start at the beginning of the line and can see here that sample batch returns a tuple of 3 things and you only use the final thing

obvs if you do get it to return 2 things then it'd be more like

```
states, _ = sample_batch(...)
return states[-1, :]
```

Reply:
thanks - that's a very useful tip